### DISCUSSION TITLE
Automating Tests

### DISCUSSION THREAD

**Discussion by PabloLec** [Textualize/textual#2506] [General]
I am currently working on an application using Textual and I am wondering how to test my application automatically, ideally in a CI environment.
I want to ensure that my application launches correctly, that navigation between different screens works, and that all screen elements are present, functional, and contain the correct data.
However, I am not sure how to do this without extreme mocking and how to programmatically manipulate the application. 
 
Has anyone done this before? Are there any ideas or examples available? Also, I would like to know if Textual has any future plans to make the testing process easier.

**Comment by davep** [comment]
Have you had a look at Textual's own tests?

**Reply by PabloLec** [reply]
No, I hadn't thought about that, but thank you for the suggestion. I'll definitely take a look!

**Comment by willmcgugan** [comment]
It's not documented at a high level, but internally we use [run_test](https://textual.textualize.io/api/app/#textual.app.App.run_test) which runs the app in headless mode. You interact with the app with a [Pilot](https://textual.textualize.io/api/pilot/#textual.pilot.Pilot) object to automate keys / clicks, then check the state with `query` and friends.

**Reply by PabloLec** [reply]
Thank you for the tip! I must admit that I overlooked the Pilot object in the documentation, but it's exactly what I needed - the little magic object that saves the day :) Thank you again for your help
